pkg/services/govmomi/service.go

```diff
@@ -218,25 +222,18 @@ func (vms *VMService) DestroyVM(ctx *context.VMContext) (infrav1.VirtualMachine,
 		State:     &vm,
 	}
 
-	// Power off the VM.
-	powerState, err := vms.getPowerState(vmCtx)
+	// Shut down the VM
+	poweredOff, err := vms.shutdownVM(vmCtx)
```

Comment:
Thanks for the comment.
I agree that it does make more sense to move the logic to the controller side to make the logic clearer.

Reply:
The logic was simplified to requeue the request when we have a pending soft power off operation, otherwise we perform power off.

We are checking the following conditions:
1. When the mode is set to `hard`, it will return `false` and a hard power off operation is done.
2. When there is a `GuestSoftPowerOffCondition`, the VM guest shutdown / soft power off has been triggered before and it's pending, unless the operation times out and the mode is set to `trySoft`.
3. If the VM never triggers guest shutdown before, we'll do it when VMware Tools is installed and the VM reports `guestStateChangeSupported`

`DestroyVM` will requeue the resource and reconcile it in 1 minute later until the guest VM is powered off.